plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs

```diff
@@ -0,0 +1,409 @@
+    , modelSubtractedSizesSlope     :: CostingInteger
+    , modelSubtractedSizesMinimum   :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+data ModelLinearSize = ModelLinearSize
+    { modelLinearSizeIntercept :: CostingInteger
+    , modelLinearSizeSlope     :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+-- | s * (x * y) + I
+data ModelMultipliedSizes = ModelMultipliedSizes
+    { modelMultipliedSizesIntercept :: CostingInteger
+    , modelMultipliedSizesSlope     :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+-- | s * min(x, y) + I
+data ModelMinSize = ModelMinSize
+    { modelMinSizeIntercept :: CostingInteger
+    , modelMinSizeSlope     :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+-- | s * max(x, y) + I
+data ModelMaxSize = ModelMaxSize
+    { modelMaxSizeIntercept :: CostingInteger
+    , modelMaxSizeSlope     :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+-- | if p then s*x else c; p depends on usage
+data ModelConstantOrLinear = ModelConstantOrLinear
+    { modelConstantOrLinearConstant  :: CostingInteger
+    , modelConstantOrLinearIntercept :: CostingInteger
+    , modelConstantOrLinearSlope     :: CostingInteger
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+-- | if p then f(x,y) else c; p depends on usage
+data ModelConstantOrTwoArguments = ModelConstantOrTwoArguments
+    { modelConstantOrTwoArgumentsConstant :: CostingInteger
+    , modelConstantOrTwoArgumentsModel    :: ModelTwoArguments
+    } deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+
+data ModelTwoArguments =
+    ModelTwoArgumentsConstantCost       CostingInteger
+  | ModelTwoArgumentsLinearInX          ModelLinearSize
+  | ModelTwoArgumentsLinearInY          ModelLinearSize
+  | ModelTwoArgumentsAddedSizes         ModelAddedSizes
+  | ModelTwoArgumentsSubtractedSizes    ModelSubtractedSizes
+  | ModelTwoArgumentsMultipliedSizes    ModelMultipliedSizes
+  | ModelTwoArgumentsMinSize            ModelMinSize
+  | ModelTwoArgumentsMaxSize            ModelMaxSize
+  | ModelTwoArgumentsLinearOnDiagonal   ModelConstantOrLinear
+  | ModelTwoArgumentsConstAboveDiagonal ModelConstantOrTwoArguments
+  | ModelTwoArgumentsConstBelowDiagonal ModelConstantOrTwoArguments
+    deriving stock (Show, Eq, Generic, Lift)
+    deriving anyclass (NFData)
+
+instance Default ModelTwoArguments where
+    def = ModelTwoArgumentsConstantCost 0
+
+runCostingFunTwoArguments
+    :: CostingFun ModelTwoArguments
+    -> ExMemory
+    -> ExMemory
+    -> ExBudget
+runCostingFunTwoArguments (CostingFun cpu mem) =
+    let !runCpu = runTwoArgumentModel cpu
+        !runMem = runTwoArgumentModel mem
+    in lazy $ \mem1 mem2 ->
```

Comment:
Used `case` instead of `let` (suggested by Heinrich Apfelmus) and we don't need that `lazy` nonsense in there now. We still need some elsewhere though.